packages/block-editor/src/store/selectors.js

```diff
@@ -1734,6 +1735,40 @@ export const __experimentalGetAllowedBlocks = createSelector(
 	]
 );
 
+/**
+ * Returns the list of allowed patterns for inner blocks children
+ *
+ * @param {Object}  state        Editor state.
+ * @param {?string} rootClientId Optional target root client ID.
+ *
+ * @return {Array?} The list of allowed block types.
+ */
+export const __experimentalGetAllowedPatterns = createSelector(
+	( state, rootClientId = null ) => {
+		const patterns = state.settings.__experimentalBlockPatterns;
+
+		if ( ! rootClientId ) {
+			return patterns;
+		}
+
+		const patternsAllowed = filter( patterns, ( { content } ) => {
+			const blocks = parse( content );
+			return every( blocks, ( { name } ) =>
+				canInsertBlockType( state, name, rootClientId )
+			);
+		} );
```

Comment:
I wonder if checking every block of every pattern could be a bit cumbersome if we are running a large amount of available patterns? 🤔   While it seems the most accurate way to handle this decision, I wonder if just whitelisting the blocks that should allow patterns as children would make more sense as it would be a much more simple check.  Is there a need to check every block, or do all parent blocks that should allow patterns support all the same children used in patterns?

Reply:
Yep, I was thinking about how we could make this better perf-wise. `createSelector` takes care of memorizing, but parsing the content of every pattern every time we select a different block is definitely cumbersome. Note, the check here only checks for the first level of blocks which means inner blocks aren't checked. Which makes sense to me.

So far my idea is to precompute the allowed patterns for each block. Then we can save it somewhere and just load it when the block editor is opened.

I'm not sure yet 🤔

Reply:
Also, reworking the selector to memoize by block type rather than client id should be a great save as well.